### PR TITLE
Define UniFFI UDL

### PR DESCRIPTION
**Pull request by mathiaspeters**
### Description
This PR is the first introduction of UniFFI into libtelio, but so far UniFFI isn't actually used. That comes in a later PR. 

With UniFFI we can make changes to our FFI code that is incompatible with SWIG and do general cleanup, so UniFFI is a breaking change, but this PR doesn't actually break anything yet. Instead it duplicates out FFI code and then cleans up the duplicate, but fear not, the duplication is not going to be long-lived.

### :ballot_box_with_check: Definition of Done checklist
- [x] Commit history is clean ([requirements](../blob/main/docs/git_commit_messages_requirements.md))
- [x] README.md is updated
- [x] changelog.md is updated
- [x] Functionality is covered by unit or integration tests

### CODE REVIEW

File: crates/telio-model/src/api_config.rs [hunk 1]
```diff
@@ -164,7 +164,7 @@ pub struct FeatureDirect {
     /// Endpoint providers [default all]
     #[serde(default)]
     #[serde(deserialize_with = "deserialize_providers")]
-    pub providers: Option<HashSet<EndpointProvider>>,
+    pub providers: Option<Vec<EndpointProvider>>,
```
Comment (mathiaspeters) [inline crates/telio-model/src/api_config.rs:167]: UniFFI can't deal with hash sets so replaced with vector

File: crates/telio-model/src/api_config.rs [hunk 2]
```diff
@@ -219,12 +219,16 @@ pub struct Features {
     pub direct: Option<FeatureDirect>,
     /// Test environment (natlab) requires binding feature disabled
     /// TODO: Remove it once mac integration tests support the binding mechanism
-    #[cfg(any(target_os = "macos", feature = "pretend_to_be_macos"))]
```
Comment (mathiaspeters) [inline crates/telio-model/src/api_config.rs:222]: UniFFI doesn't like conditionally including/excluding stuff it exposes so this field will now always be there

File: crates/telio-model/src/api_config.rs [hunk 3]
```diff
@@ -219,12 +219,16 @@ pub struct Features {
...
+    #[serde(default = "default_is_test_env")]
     pub is_test_env: Option<bool>,
     /// Derp server specific configuration
     pub derp: Option<FeatureDerp>,
 }
 
+fn default_is_test_env() -> Option<bool> {
+    Some(true)
+}
```
Comment (mathiaspeters) [inline crates/telio-model/src/api_config.rs:230]: I don't fully remember why this defaults to true, but I remember trying to just do the type default (back then false, because it wasn't yet an option) and that broke stuff in tests

File: src/ffi/types.rs [hunk 4]
```diff
@@ -3,10 +3,111 @@ use log::Level;
...
+    #[error("LockError")]
+    LockError,
+    #[error("InvalidString")]
+    InvalidString,
+    #[error("AlreadyStarted")]
+    AlreadyStarted,
+}
+
+#[derive(Copy, Clone, Debug)]
+#[allow(clippy::enum_variant_names)] // Otherwise clippy complains about all variants having "Tun" as their suffix
```
Comment (mathiaspeters) [inline src/ffi/types.rs:43]: Without this line the compilers/linters complain that all variants end with `Tun`

File: src/libtelio.udl [hunk 5]
```diff
@@ -0,0 +1,520 @@
...
+    /// # Parameters
+    /// - `log_level`: Max log level to log.
+    /// - `logger`: Callback to handle logging events.
+    void set_global_logger(TelioLogLevel log_level, TelioLoggerCb logger);
+
+    /// Get default recommended adapter type for platform.
+    TelioAdapterType uniffi_telio_get_default_adapter();
+
+    /// Get current version tag.
+    string uniffi_telio_get_version_tag();
```
Comment (mathiaspeters) [inline src/libtelio.udl:30]: Some functions here are prefixed with `uniffi_telio_`, that is temporary and will be removed in a future PR. It's just to prevent name clashes with existing functions until those functions are removed

File: src/ffi/mod.rs [hunk 6]
```diff
@@ -82,9 +113,567 @@ lazy_static::lazy_static! {
...
+        error!("{}", err)
+    }
+
+    log::set_max_level(
+        (<types::TelioLogLevel as Into<Level>>::into(log_level) as Level).to_level_filter(),
+    );
+}
+
+/// Get default recommended adapter type for platform.
+pub fn uniffi_telio_get_default_adapter() -> TelioAdapterType {
```
Comment (mathiaspeters) [inline src/ffi/mod.rs:131]: See my other comment about temporary name prefixes

**Comment by tomaszklak** [comment]
Can you add in this PR something like `ffi/tests` to prove that this is at least working a little bit? As far as I understand this PR lacks any form of testing/verification.

File: Cargo.toml [hunk 7]
```diff
@@ -38,6 +38,7 @@ tokio = { version = ">=1.22", features = ["full"] }
 async-trait = "0.1.51"
 uuid =  { version = "1.1.2", features = ["v4"] }
 wasm-bindgen = "=0.2.83" # Only for compatability with moose wasm version
+uniffi.workspace = true
```
[the remark this answers is not in the document]
Comment (jjanowsk) [reply Cargo.toml:41]: Also as I understand this is done just to remove duplication of `uniffi = { git = "https://github.com/NordSecurity/uniffi-rs.git", branch = "nordsec.0.23.0" }` but I would advise to use official uniffi 0.23.0 so maybe it's not a problem to just have
```
uniffi = "0.23.0"
...
uniffi = { version = "0.23.0", features = ["build"] }
```
I know it still requires updates in two places for version bumps but reduces the complexity of the toml in general. 

Just my 2 cents

[the remark this answers is not in the document]
Comment (mathiaspeters) [reply Cargo.toml:41]: As far as I know, upstream uniffi doesn't support doc comments so probably best to stick to our fork. But I can duplicate the version information

[the remark this answers is not in the document]
Comment (jjanowsk) [reply Cargo.toml:41]: Only generators must support docstrings. Scaffolding part does not need to know about it and I believe upstream just ignores comments (this is to be checked) so it should work with this udl. And we keep compatibility between upstream and our fork. So you should be able to safely use upstream 0.23.0 and use our generators to support docstrings in the generated bindings

[the remark this answers is not in the document]
Comment (mathiaspeters) [reply Cargo.toml:41]: Ah, there's been a misunderstanding then. I asked about this in a slack thread and was told both the rust dependency and the generator needs to be our fork for doc comments to work, but it was probably me being unclear in my question. Good to know we can use upstream here

File: crates/telio-model/src/api_config.rs (hunk 3 above)
Comment (mathiaspeters) [reply crates/telio-model/src/api_config.rs:230]: I can do that, though that `cfg` attribute doesn't persist across crate boundaries so other crates who depend on `telio-model` would get the `not(test)` branch all the time

Comment (jjanowsk) [reply crates/telio-model/src/api_config.rs:230]: IMO it should default to false. In current form it will be defaulted to true also in production which is incorrect. It should default to false and be changed to true only via config during natlab tests

Comment (mathiaspeters) [reply crates/telio-model/src/api_config.rs:230]: I don't know why it became an optional in the first place, but opinion of defaulting to `None`?

Comment (jjanowsk) [reply crates/telio-model/src/api_config.rs:230]: It should also be fine, and maybe even the cleanest option

File: crates/telio-model/src/config.rs [hunk 8]
```diff
@@ -194,6 +194,18 @@ pub struct Config {
     pub dns: Option<DnsConfig>,
 }
 
+impl Config {
+    /// Check if the config is empty
+    pub fn is_empty(&self) -> bool {
+        self.this.hostname.is_empty()
+            && self.this.identifier.is_empty()
+            && self.this.ip_addresses.is_none()
```
Comment (jjanowsk) [inline crates/telio-model/src/config.rs:202]: Any reason to omit: `self.this.public_key`?

Comment (tomaszklak) [reply crates/telio-model/src/config.rs:202]: Public key is non optional and it itself is wrapper over `[u8; KEY_SIZE]` so it never is empty (unless you would count something like `[0; KEY_SIZE]` as empty).

Comment (mathiaspeters) [reply crates/telio-model/src/config.rs:202]: I think it was mainly a decision of convenience, influenced by exactly what Tomasz is saying. It will never be empty so the next best thing is to check that it's all 0, but if everything else in the config is empty, we can already say that the config is empty and skip that check. It's not the cleanest solution but it was the most convenient, and one made to keep the interface more similar. Though it's probably better, now that we don't deserialize strings into configs to not have the clients send None instead of an empty config if that's what they mean. After all, the clients should now better in which scenarios the config is empty by intention

**Comment by tomaszklak** [comment]
I've checked the generated code and it looks like it solve the issue we had with using `repr(C)` for enums that @stalowyjez tried to solve here: https://bucket.digitalarsenal.net/low-level-hacks/vpn/client/libtelio-oss/-/merge_requests/144

Could you verify this by adding a test (i think it would have to be a C test) that tries to call new ffi functions with enum values out of range?

File: src/ffi/types.rs [hunk 9]
```diff
@@ -231,6 +332,36 @@ map_enum! {
...
+    Error = Error,
+    Warn = Warning,
+    Info = Info,
+    Debug = Debug,
+    Trace = Trace,
+}
+
+map_enum! {
+    TelioLogLevel -> Level,
+    Critical = Error,
```
Comment (jjanowsk) [inline src/ffi/types.rs:357]: Can we drop Critical from our interface since we modify it anyway?

Comment (mathiaspeters) [reply src/ffi/types.rs]: Done

File: Cargo.toml (hunk 7 above)
[the remark this answers is not in the document]
Comment (mathiaspeters) [reply Cargo.toml]: Done

File: crates/telio-model/src/api_config.rs [hunk 10]
```diff
@@ -187,7 +187,7 @@ where
         Err(_) => return Ok(None),
     };
 
-    let eps: HashSet<_> = eps
+    let eps: Vec<_> = eps
```
[the remark this answers is not in the document]
Comment (mathiaspeters) [reply crates/telio-model/src/api_config.rs:202]: This deserializer block might actually be possible to remove at some point, since we're no longer getting and deserializing strings for domain objects. I added deduplication of endpoint providers to the constructor of `Device`

File: crates/telio-model/src/api_config.rs (hunk 1 above)
Comment (mathiaspeters) [reply crates/telio-model/src/api_config.rs:169]: I tried using a type alias for `HashSet<EndpointProvider>` and I tried creating a wrapper, and neither worked. I think it should be possible to have a wrapper but so far I feel like it's much more work than it's worth given how little we actually use the endpoint providers. Instead I added deduplication of endpoint providers to the constructor for `Device`

**Comment by mathiaspeters** [comment]
Regarding adding tests for the FFI bindings, as mentioned by Tomasz, in this PR:
When we're done migrating to UniFFI, all official apps will be using high-level languages to interact with libtelio, and the interface will be strongly typed, so adding tests for things like calling a function with an invalid enum value does not make that much sense. As for adding more general tests, a follow-up PR will be migrating the natlab tests to use these bindings instead of `tcli`, and these bindings will not be used before then, so adding dedicated FFI tests like in `ffi/tests` would also not bring much value.

**Review by tomaszklak — APPROVED** [review]
+1

File: crates/telio-model/src/api_config.rs [hunk 11]
```diff
@@ -179,16 +191,21 @@ pub struct FeatureDerp {
 }
 
 /// Whether to validate keys
-#[derive(Clone, Debug, PartialEq, Eq, Deserialize)]
-pub struct FeatureValidateKeys(pub bool);
+#[derive(Copy, Clone, Debug, PartialEq, Eq, Deserialize)]
+pub struct FeatureValidateKeys {
+    /// Defaults to true
+    pub validate_keys: bool,
+}
```
Comment (mathiaspeters) [inline crates/telio-model/src/api_config.rs:198]: Reminder for myself: this change will require a change in the confluence page for libtelio features

File: crates/telio-model/src/api_config.rs [hunk 12]
```diff
@@ -244,19 +249,27 @@ pub struct FeatureDerp {
...
 }
 
 /// Turns on connection resets upon VPN server change
 #[derive(Default, Clone, Copy, Debug, PartialEq, Eq, Deserialize)]
 #[serde(transparent)]
-pub struct FeatureBoringtunResetConns(pub bool);
+pub struct FeatureBoringtunResetConns {
+    /// Defults to false
+    pub reset_connections: bool,
+}
```
Comment (mathiaspeters) [inline crates/telio-model/src/api_config.rs:272]: Reminder for myself: this change will require a change in the confluence page for libtelio features

File: tests/uniffi_logger.rs [hunk 13]
```diff
@@ -0,0 +1,42 @@
...
+        atomic::{AtomicUsize, Ordering},
+        Arc,
+    };
+
+    use telio::ffi_types::TelioLoggerCb;
+
+    use super::*;
+
+    #[test]
+    fn uniffi_test_logger() {
```
Comment (mathiaspeters) [inline tests/uniffi_logger.rs:15]: `tracing::set_global_default` is necessary to make this test work, but there can't be called twice in the same test file, that's why this test got it's own file instead of also placing it in `logger.rs`. The duplication will be fixed in a later PR

File: crates/telio-model/src/api_config.rs (hunk 1 above)
Comment (mathiaspeters) [reply crates/telio-model/src/api_config.rs]: I managed to fix it with a type alias

File: crates/telio-model/src/api_config.rs (hunk 10 above)
[the remark this answers is not in the document]
Comment (mathiaspeters) [reply crates/telio-model/src/api_config.rs]: No longer relevant

File: crates/telio-model/src/api_config.rs (hunk 12 above)
Comment (tomaszklak) [reply crates/telio-model/src/api_config.rs:270]: why is that? It seems that the json format is the same 🤔

File: crates/telio-model/src/api_config.rs (hunk 11 above)
Comment (tomaszklak) [reply crates/telio-model/src/api_config.rs:254]: I think if you add `repr(transparent)` you will keep the json format unchanged.

File: src/device/mod.rs [hunk 14]
```diff
@@ -2728,10 +2728,13 @@ mod tests {
 
         let (sender, _receiver) = tokio::sync::broadcast::channel(1);
 
-        let mut providers = HashSet::<telio_model::api_config::EndpointProvider>::new();
-        providers.insert(telio_model::api_config::EndpointProvider::Stun);
-        providers.insert(telio_model::api_config::EndpointProvider::Upnp);
-        providers.insert(telio_model::api_config::EndpointProvider::Local);
+        let providers = vec![
```
Comment (tomaszklak) [inline src/device/mod.rs:2731]: nit: you don't need a vector, an array should also work but you can use `maplit::hashset` instead of creating temporary collection.

File: src/ffi/mod.rs [hunk 15]
```diff
@@ -85,12 +119,604 @@ lazy_static::lazy_static! {
...
+    }
+
+    /// Explicitly deallocate telio object and shutdown async rt.
+    pub fn destroy_hard(&self) -> FFIResult<()> {
+        let mut dev = self
+            .inner
+            .lock()
+            .unwrap_or_else(|poisoned| poisoned.into_inner());
+        let res = catch_ffi_panic(|| {
+            dev.try_shutdown(Duration::from_millis(1000))
```
Comment (tomaszklak) [inline src/ffi/mod.rs:303]: Previously `try_shutdown` was destroying `dev` here (and thus deallocating whole object tree starting in `dev`). Why is this no longer done?

File: src/ffi/mod.rs [hunk 16]
```diff
@@ -1063,8 +1689,70 @@ impl Subscriber for TelioTracingSubscriber {
     }
 }
 
+pub struct UniffiTelioTracingSubscriber {
```
Comment (tomaszklak) [inline src/ffi/mod.rs:1692]: Why is this a copy pasted version of `TelioTracingSubscriber`?

File: src/ffi/mod.rs [hunk 17]
```diff
@@ -85,12 +119,604 @@ lazy_static::lazy_static! {
...
+                        .set(format!("{}", info)),
+                );
+
+                telio_log_debug!("call_once: {:?}", e);
+                events(e);
+            }));
+        });
+
+        // TODO: Update windows ffi to take in void*, for protect
+        #[cfg(not(target_os = "android"))]
```
Comment (packgron) [inline src/ffi/mod.rs:250]: Do we really need this to be gatted under android. I think the option is universal enough, and in the future someone needs to customize this, I see no problem with that.

File: src/ffi/mod.rs [hunk 18]
```diff
@@ -85,12 +119,604 @@ lazy_static::lazy_static! {
...
+        catch_ffi_panic(|| {
+            dev.stop();
+            Ok(())
+        })
+    }
+
+    /// get device luid.
+    pub fn get_adapter_luid(&self) -> u64 {
+        match self.inner.lock() {
+            Ok(mut d) => d.get_adapter_luid(),
```
Comment (packgron) [inline src/ffi/mod.rs:427]: What will be returned for unix platforms ?

File: src/lib.rs [hunk 19]
```diff
@@ -51,3 +53,124 @@ pub use telio_lana;
...
+
+    impl UniffiCustomTypeConverter for PublicKey {
+        type Builtin = String;
+
+        fn into_custom(val: Self::Builtin) -> uniffi::Result<Self> {
+            Ok(val.parse().map_err(|_| TelioError::InvalidKey)?)
+        }
+
+        fn from_custom(obj: Self) -> Self::Builtin {
+            String::from_utf8_lossy(&obj.0).to_string()
```
Comment (tomaszklak) [inline src/lib.rs:80]: I don't think this is correct, why not instead convert it to something like base64?

File: src/lib.rs [hunk 20]
```diff
@@ -51,3 +53,124 @@ pub use telio_lana;
 
 /// cbindgen:ignore
 pub use telio_firewall;
+
+pub use uniffi_libtelio::*;
+#[allow(clippy::panic, clippy::unwrap_used, clippy::expect_used, unwrap_check)]
+mod uniffi_libtelio {
```
Comment (tomaszklak) [inline src/lib.rs:59]: All of the conversion here should have tests.

File: src/libtelio.udl [hunk 21]
```diff
@@ -0,0 +1,572 @@
...
+    /// Derp will send empty messages after this many seconds of not sending/receiving any data [default 60s]
+    u32? derp_keepalive;
+    /// Enable polling of remote peer states to reduce derp traffic
+    boolean? enable_polling;
+    /// Use Mozilla's root certificates instead of OS ones [default false]
+    boolean use_built_in_root_certificates;
+};
+
+/// Whether to validate keys
+dictionary FeatureValidateKeys {
```
Comment (packgron) [inline src/libtelio.udl:450]: maybe this just be a boolean?

File: src/libtelio.udl [hunk 22]
```diff
@@ -0,0 +1,572 @@
...
+    /// Defaults to true
+    boolean validate_keys;
+};
+
+/// Turns on connection resets upon VPN server change
+dictionary FeatureBoringtunResetConns {
```
Comment (packgron) [inline src/libtelio.udl:456]: maybe this should just be a boolean?

File: src/libtelio.udl [hunk 23]
```diff
@@ -0,0 +1,572 @@
...
+callback interface TelioLoggerCb {
+    void log(TelioLogLevel log_level, string payload);
+};
+
+callback interface TelioProtectCb {
+    void protect(i32 payload);
+};
+
+/// Encompasses all of the possible features that can be enabled
+dictionary Features {
```
[the remark this answers is not in the document]
Comment (tomaszklak) [reply src/libtelio.udl:322]: I imagine that this can make going back and forth between libtelio versions slightly harder, but that is mostly a concern for the apps dev teams so I think they should also be involved here.

File: src/ffi/mod.rs (hunk 16 above)
Comment (mathiaspeters) [reply src/ffi/mod.rs:1692]: Because the old code is still alive, but the new code should still be tested. There will be a cleanup PR later where all of the duplication introduced here is removed (the `uniffi_` and `UniffI` prefixes will be removed at the same time)

File: src/ffi/mod.rs (hunk 15 above)
Comment (mathiaspeters) [reply src/ffi/mod.rs:303]: Because UniFFI doesn't allow you to take ownership of self. It doesn't even allow you to take a mutable reference to self

File: src/ffi/mod.rs (hunk 17 above)
Comment (mathiaspeters) [reply src/ffi/mod.rs:250]: I was following the old logic here, but I'm happy to not conditionally compile this

File: src/ffi/mod.rs (hunk 18 above)
Comment (mathiaspeters) [reply src/ffi/mod.rs:427]: `boringtun` and `linux-native` are hard coded to return 0 in `boring.rs` and `linux_native_wg.rs`

File: src/ffi/mod.rs (hunk 16 above)
Comment (tomaszklak) [reply src/ffi/mod.rs:1692]: I think that if you change the `TelioTracingSubscriber` to also use `TelioLoggerCb` there will be no need for `UniffiTelioTracingSubscriber`.

File: src/ffi/mod.rs (hunk 15 above)
Comment (tomaszklak) [reply src/ffi/mod.rs:303]: I understand, but why are we no longer destroying that objects/memory - is that still correct behaviour? Maybe `telio` should store represent `inner` as `Mutex<Option<Device>>` and set it to `None` as part of the `try_shutdown`?

Comment (mathiaspeters) [reply src/ffi/mod.rs:303]: That sounds like a valid option. I will make the change

File: src/libtelio.udl (hunk 22 above)
Comment (mathiaspeters) [reply src/libtelio.udl]: Yes, I updated these to not have to do custom conversion logic for UniFFI but of course it's better to keep the API simple than to avoid such conversions. Fixed

File: src/libtelio.udl (hunk 21 above)
Comment (mathiaspeters) [reply src/libtelio.udl]: Yes, I updated these to not have to do custom conversion logic for UniFFI but of course it's better to keep the API simple than to avoid such conversions. Fixed

File: src/lib.rs (hunk 20 above)
Comment (mathiaspeters) [reply src/lib.rs:59]: Added

File: src/lib.rs (hunk 19 above)
Comment (mathiaspeters) [reply src/lib.rs]: I updated these to use `Vec<u8>`instead of `String` since that is what the keys use under the hood. I think I remember getting some errors when I tried to do that in the past, but it worked now

File: src/ffi/mod.rs (hunk 16 above)
Comment (mathiaspeters) [reply src/ffi/mod.rs]: I added an enum to wrap the callback

File: src/device/mod.rs (hunk 14 above)
Comment (mathiaspeters) [reply src/device/mod.rs]: Fixed

File: crates/telio-model/src/api_config.rs (hunk 12 above)
Comment (mathiaspeters) [reply crates/telio-model/src/api_config.rs]: Yes, I updated these to not have to do custom conversion logic for UniFFI but of course it's better to keep the API simple than to avoid such conversions. Fixed

File: crates/telio-model/src/api_config.rs (hunk 11 above)
Comment (mathiaspeters) [reply crates/telio-model/src/api_config.rs]: Yes, I updated these to not have to do custom conversion logic for UniFFI but of course it's better to keep the API simple than to avoid such conversions. Fixed

File: src/ffi/mod.rs (hunk 17 above)
Comment (mathiaspeters) [reply src/ffi/mod.rs]: It seems like file descriptors are u64 on windows but i32 on all other platforms. I can either try to dig into the code and add a conversion wherever windows is calling it, or a simpler option, exclude the protect closure on windows. Opinion?

**Review by packgron — APPROVED** [review]
+1.0

File: src/ffi/mod.rs (hunk 17 above)
Comment (packgron) [reply src/ffi/mod.rs]: so let's this this for unix only

**Review by packgron — APPROVED** [review]
